### ISSUE TITLE
Script only passes one identifier per UserData object

### ISSUE THREAD

**Issue by jradcliff** [google/customer-match-upload-script#2]
It appears that [`build_offline_user_data_job_operations`](https://github.com/google/customer-match-upload-script/blob/master/create_and_populate_list.py#L330) creates a new `OfflineUserDataJobOperation` and `UserData` object for _each_ `data_type` found in a row. If a single row in the CSV contains entries for multiple data types such as `emails` and `addresses`, you should create a _single_ `UserData` with multiple `user_identifiers` instead.

So instead of two separate UserData (`create`) objects for the same member of the list (row in the CSV):

```
operations {
  create {
    user_identifiers {
      address_info {
        hashed_first_name: "<hashed first name here>"
        hashed_last_name: "<hashed last name here>"
        country_code: "<country code>"
        postal_code: "<postal code>"
      }
    }
  }
  create {
    user_identifiers {
      hashed_email: "<hashed email address here>"
    }
  }
}
```

You should have _one_ UserData with multiple `user_identifiers`:

```
operations {
  create {
    user_identifiers {
      hashed_email: "<hashed email address here>"
    }
    user_identifiers {
      address_info {
        hashed_first_name: "<hashed first name here>"
        hashed_last_name: "<hashed last name here>"
        country_code: "<country code>"
        postal_code: "<postal code>"
      }
    }
  }
}
```

You still want a separate UserData for each distinct _member_ of the list, but all identifiers for the same member should go under one UserData.

Hope that's helpful!

**Comment by dliu9999** [comment]
Does this change significantly impact performance at all? I'm trying to upload about 1.4million rows, each row with [Email, Phone, FirstName, LastName, CountryCode, ZipCode], with the existing script, but `build_offline_user_data_job_operations` took over 1.5 hrs before I killed it.

Do you have any other suggestions?

**Comment by jradcliff** [comment]
Hi,

I would caution against extrapolating job run times from such a small sample, as there's overhead involved in every job and jobs are queued up before they run. See the first bullet point [in our guide](https://developers.google.com/google-ads/api/docs/remarketing/audience-types/customer-match#customer_match_considerations) for expected job run times.

Even if using a separate `UserData` with multiple identifiers for each distinct member results in longer processing times, you should still use that approach since that's the correct usage.

Thanks,
Josh

**Comment by valerioditinno** [comment]
hello @[jradcliff](https://github.com/jradcliff),  

I had your same concern and I basically refactored **read_csv** and **build_offline_user_data_job_operations** in order to send exactly one request to the api for each row in the csv. 
My question is: are you sure that this won't impact the overall match rate? 
Did you get in touch with someone from google side that validate your idea?

Thanks
Valerio

**Comment by valerioditinno** [comment]
Yes, you are right I meant operations and sorry I had no idea you were from the Google Ads API team :)

I have a last doubt: why I need to create differents user_identifiers for the same member of the list?

So basically why this:

```
operations {
  create {
    user_identifiers {
      hashed_email: "<hashed email address here>"
    }
    user_identifiers {
      hashed_phone_number: "<hashed phone number address here>"
    }
    user_identifiers {
      address_info {
        hashed_first_name: "<hashed first name here>"
        hashed_last_name: "<hashed last name here>"
        country_code: "<country code>"
        postal_code: "<postal code>"
      }
    }
  }
}
```
instead of this:

```
operations {
  create {
    user_identifiers {
      hashed_email: "<hashed email address here>"
      hashed_phone_number: "<hashed phone number address here>"
      address_info {
        hashed_first_name: "<hashed first name here>"
        hashed_last_name: "<hashed last name here>"
        country_code: "<country code>"
        postal_code: "<postal code>"
      }
    }
  }
}
```

Thank you for your time

**Comment by valedtn** [comment]
Hi @ashishghai, sorry for the late response.

I send the data to the API using an Apache Beam pipeline (Python SDK) that use BatchElements PTransform: this transform attempts to optimize the batch size based on total elements in the PCollection, so I do not have a fixed one. 
For 15k elements the pipeline creates 3 batch and it takes 7 min and 20 sec to complete (batch are processed sequentially)

Hope it helps a little

Have a nice day

Valerio

**Comment by shreesmart** [comment]
Hi @valerioditinno Thank you very much !! it gives idea about latency . you too have a great day 

Kind regards,
Ashish

**Comment by dliu9999** [comment]
Hi,

Understandable, but this is only building the offline data object locally,
not running it async yet. Also I tested on sizes of 10, 50, and 100, and
they scaled linearly (5, 25, 50 seconds respectively).

So you think this is still the only/correct way? Some more context: we want
to be able to update the records with changes (either removals or
additions) once a week, I think currently we are just uploading the full
updated csv each week so we're thinking about doing the same but with
python.

**Comment by RyHoLi** [comment]
Hi @jradcliff or @valerioditinno,

How can we update the script so that we can get one UserData with multiple identifiers instead of two separate UserData objects for the same member of the list? I believe this is causing me to get an error that says "Maximum number of user identifiers allowed per request is 100,000 and per operation is 20."

**Comment by jradcliff** [comment]
Hi,

If you have multiple user identifiers for the same UserData object, you would call `user_data.user_identifiers.append(...)` multiple times. For example:

```
...
customer_data_operations = []
...

// Creates the operation and UserData objects.
user_data_operation = client.get_type('OfflineUserDataJobOperation')
user_data = user_data_operation.create

// Creates and adds identifier for hashed email to the UserData.
email_user_identifier = client.get_type('UserIdentifier')
email_user_identifier.hashed_email = item['hashed_email']
user_data.user_identifiers.append(email_user_identifier)

// Creates and adds identifier for hashed phone to the same UserData.
phone_user_identifier = client.get_type('UserIdentifier') 
phone_user_identifier.hashed_phone_number = item['hashed_phone_number']
user_data.user_identifiers.append(phone_user_identifier)

// Adds the ONE operation for the UserData to the collection.
customer_data_operations.append(user_data_operation)
```

The error `Maximum number of user identifiers allowed per request is 100,000 and per operation is 20` occurs if either of the following conditions holds:

1. The _total_ number of `UserIdentifier` objects across _all_ operations in the request exceeds 100,000.
  - In this case, you can just split the operations across multiple requests.
2. A _single_ operation's `UserData` contains more than 20 elements in its `user_identifiers` collection.
  - This usually indicates an implementation error in how you are constructing your `UserData` objects, as it's unlikely you would have more than 20 identifiers for a single user.

See our blog post at https://ads-developers.googleblog.com/2021/10/userdata-enforcement-in-google-ads-api.html for more details.

Thanks,
Josh, Google Ads API Team
